src/differentiation/vecjac_products.jl

```diff
@@ -47,31 +51,121 @@ mul!(w, L, v) # = df/du * v
 L(v, p, t; VJP_input = w)    # = df/dw * v
 L(x, v, p, t; VJP_input = w) # = df/dw * v
 ```
+
+## Allowed Function Signatures for `f`
+
+For Out of Place Functions:
+
+```julia
+f(u, p, t)  # t !== nothing
+f(u, p)     # p !== nothing
+f(u)        # Otherwise
+```
+
+For In Place Functions:
+
+```julia
+f(du, u, p, t)  # t !== nothing
+f(du, u, p)     # p !== nothing
+f(du, u)        # Otherwise
+```
 """
-function VecJac(f, u::AbstractArray, p = nothing, t = nothing;
+function VecJac(f, u::AbstractArray, p = nothing, t = nothing; fu = nothing,
         autodiff = AutoFiniteDiff(), kwargs...)
-    L = _vecjac(f, u, autodiff)
-    IIP, OOP = get_iip_oop(L)
+    ff = VecJacFunctionWrapper(f, fu, u, p, t)
 
-    if isa(autodiff, AutoZygote) & !OOP
+    if !__internal_oop(ff) && autodiff isa AutoZygote
         msg = "Zygote requires an out of place method with signature f(u)."
         throw(ArgumentError(msg))
     end
 
-    # NOTE: The operator returned has both in-place and out-of-place definitions and
-    #       doesn't follow the convention of `f`
-    return FunctionOperator(L, u, u; isinplace = true, outofplace = OOP,
-        p, t, islinear = true, accepted_kwargs = (:VJP_input,), kwargs...)
+    fu === nothing && (fu = __internal_oop(ff) ? ff(u) : u)
+
+    op = _vecjac(ff, fu, u, autodiff)
+
+    # FIXME: FunctionOperator is terribly type unstable. It makes it `::Any`
+    # NOTE: We pass `p`, `t` to Function Operator but we always use the cached version from
+    #       VecJacFunctionWrapper
```

Comment:
Sounds like the right thing to do is fix FunctionOperator?

Reply:
the `note` part or the `fixme` part? The `fixme` part definitely needs fixing, but the `note` part is needed, because `p` and `t` need to be propagated via `update_coefficients`